### PR TITLE
Add ability to change the header format

### PR DESCRIPTION
**Pull request by cgzones**
Instead of just two equals sized columns, add the ability to select from predefined formats.

**NOTE**: Please backup your htop configuration (*~/.config/htop/htoprc*) before testing, cause after running this modified htop and reverting back to master, the configuration file becomes invalid!

### CODE REVIEW

File: Header.c [hunk 1]
```diff
@@ -149,24 +195,36 @@ void Header_reinit(Header* this) {
    }
 }
 
+static const int formatWidths[][5] = {
```
Comment (fasterit) [inline Header.c:198]: I think this should go in the Header.h file

- can be reused elsewhere
- proximity to typedef enum HeaderFormat_  {..} (which should be moved from Settings.h)

File: Settings.h [hunk 2]
```diff
@@ -17,15 +17,28 @@ in the source distribution for its full text.
 
 #define DEFAULT_DELAY 15
 
+typedef enum HeaderFormat_ {
```
Comment (fasterit) [inline Settings.h:20]: Better co-locate with the formatWidths array for easier maintainability, see comment above

File: CategoriesPanel.c [hunk 3]
```diff
@@ -139,5 +159,6 @@ CategoriesPanel* CategoriesPanel_new(ScreenManager* scr, Settings* settings, Hea
    Panel_add(super, (Object*) ListItem_new("Display options", 0));
    Panel_add(super, (Object*) ListItem_new("Colors", 0));
    Panel_add(super, (Object*) ListItem_new("Columns", 0));
+   Panel_add(super, (Object*) ListItem_new("Header Format", 0));
```
Comment (fasterit) [inline CategoriesPanel.c:162]: Header **Layout** ?
(and then s/Format/Layout/ig on the variable names / settings as well)

File: AvailableMetersPanel.h [hunk 4]
```diff
@@ -19,12 +20,12 @@ typedef struct AvailableMetersPanel_ {
 
    Settings* settings;
    Header* header;
-   Panel* leftPanel;
-   Panel* rightPanel;
+   unsigned int columns;
```
Comment (BenBE) [inline AvailableMetersPanel.h:23]: Mark countable amounts of things as `size_t`?

File: AvailableMetersPanel.h [hunk 5]
```diff
@@ -19,12 +20,12 @@ typedef struct AvailableMetersPanel_ {
...
+   MetersPanel** meterPanels;
 } AvailableMetersPanel;
 
 extern const PanelClass AvailableMetersPanel_class;
 
-AvailableMetersPanel* AvailableMetersPanel_new(Settings* settings, Header* header, Panel* leftMeters, Panel* rightMeters, ScreenManager* scr, const ProcessList* pl);
+AvailableMetersPanel* AvailableMetersPanel_new(Settings* settings, Header* header, unsigned int columns, MetersPanel **meterPanels, ScreenManager* scr, const ProcessList* pl);
```
Comment (BenBE) [inline AvailableMetersPanel.h:29]: dito.

File: CategoriesPanel.c [hunk 6]
```diff
@@ -34,13 +36,23 @@ static void CategoriesPanel_delete(Object* object) {
...
 
 void CategoriesPanel_makeMetersPage(CategoriesPanel* this) {
-   MetersPanel* leftMeters = MetersPanel_new(this->settings, "Left column", this->header->columns[0], this->scr);
-   MetersPanel* rightMeters = MetersPanel_new(this->settings, "Right column", this->header->columns[1], this->scr);
-   leftMeters->rightNeighbor = rightMeters;
-   rightMeters->leftNeighbor = leftMeters;
-   Panel* availableMeters = (Panel*) AvailableMetersPanel_new(this->settings, this->header, (Panel*) leftMeters, (Panel*) rightMeters, this->scr, this->pl);
-   ScreenManager_add(this->scr, (Panel*) leftMeters, 20);
-   ScreenManager_add(this->scr, (Panel*) rightMeters, 20);
+   unsigned int columns = HeaderLayout_getColumns(this->scr->header->headerLayout);
```
Comment (BenBE) [inline CategoriesPanel.c:39]: dito.

File: CategoriesPanel.c [hunk 7]
```diff
@@ -34,13 +36,23 @@ static void CategoriesPanel_delete(Object* object) {
...
+   MetersPanel** meterPanels = xMallocArray(columns, sizeof(MetersPanel));
+
+   for (unsigned int i = 0; i < columns; i++) {
```
Comment (BenBE) [inline CategoriesPanel.c:42]: dito.

File: ColorsPanel.h [hunk 8]
```diff
@@ -15,11 +15,10 @@ typedef struct ColorsPanel_ {
    Panel super;
 
    Settings* settings;
-   ScreenManager* scr;
 } ColorsPanel;
 
 extern const PanelClass ColorsPanel_class;
 
-ColorsPanel* ColorsPanel_new(Settings* settings, ScreenManager* scr);
```
Comment (BenBE) [inline ColorsPanel.h:23]: How much is this change related to the header format changes?

File: Header.c [hunk 9]
```diff
@@ -20,51 +21,91 @@ in the source distribution for its full text.
 #include "XUtils.h"
 
 
-Header* Header_new(struct ProcessList_* pl, Settings* settings, int nrColumns) {
-   Header* this = xCalloc(1, sizeof(Header));
-   this->columns = xCalloc(nrColumns, sizeof(Vector*));
+Header* Header_new(struct ProcessList_* pl, Settings* settings, HeaderLayout hLayout) {
+   Header* this = xMalloc(sizeof(Header));
```
Comment (BenBE) [inline Header.c:25]: Why not ensure zero-allocated structure with `xCalloc`?

File: Header.c [hunk 10]
```diff
@@ -20,51 +21,91 @@ in the source distribution for its full text.
...
       Vector_delete(this->columns[i]);
    }
+
    free(this->columns);
    free(this);
 }
 
+void Header_setLayout(Header* this, HeaderLayout hLayout) {
+   unsigned int oldColumns = HeaderLayout_getColumns(this->headerLayout);
+   unsigned int newColumns = HeaderLayout_getColumns(hLayout);
```
Comment (BenBE) [inline Header.c:49]: Counts as `size_t`?

File: Header.c [hunk 11]
```diff
@@ -20,51 +21,91 @@ in the source distribution for its full text.
...
+
 void Header_populateFromSettings(Header* this) {
+   Header_setLayout(this, this->settings->hLayout);
+
    Header_forEachColumn(this, col) {
-      const MeterColumnSettings* colSettings = &this->settings->columns[col];
-      for (int i = 0; i < colSettings->len; i++) {
+      const MeterColumnSetting* colSettings = &this->settings->hColumns[col];
+      Vector_prune(this->columns[col]);
+      for (uint8_t i = 0; i < colSettings->len; i++) {
```
Comment (BenBE) [inline Header.c:80]: Risk of infinite loop if `colSettings->len>=256`

File: Header.h [hunk 12]
```diff
@@ -12,34 +12,37 @@ in the source distribution for its full text.
...
    Vector** columns;
    Settings* settings;
    ProcessList* pl;
-   int nrColumns;
+   HeaderLayout headerLayout;
    int pad;
    int height;
 } Header;
 
-#define Header_forEachColumn(this_, i_) for (int (i_)=0; (i_) < (this_)->nrColumns; ++(i_))
```
Comment (BenBE) [inline Header.h:24]: Use `size_t` for countables …

File: HeaderLayout.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+typedef struct HeaderLayoutImpl_ {
+   HeaderLayout id;
+   unsigned int columns;
+   const unsigned int widths[4];
+   const char* description;
+} HeaderLayoutImpl;
+
+static const HeaderLayoutImpl HeaderLayout_layouts[LAST_HEADER_LAYOUT] = {
+   { HF_TWO_50_50,        2, { 50, 50,  0,  0 }, "2 columns - 50/50 (default)", },
```
Comment (BenBE) [inline HeaderLayout.h:36]: Why not reuse a similar array index structure as with the color schemes?

```c
  [HF_TWO_50_50] = {2, { 50, 50, 0, 0}, "2 columns - 50/50 (default)" },
  ...
```

File: HeaderOptionsPanel.c [hunk 14]
```diff
@@ -0,0 +1,87 @@
+/*
+htop - HeaderOptionsPanel.c
+(C) 2020 htop dev team
```
Comment (BenBE) [inline HeaderOptionsPanel.c:3]: Shouldn't this be 2020-2021?

File: HeaderOptionsPanel.h [hunk 15]
```diff
@@ -0,0 +1,25 @@
+#ifndef HEADER_HeaderOptionsPanel
+#define HEADER_HeaderOptionsPanel
+/*
+htop - ColorsPanel.h
+(C) 2004-2011 Hisham H. Muhammad
```
Comment (BenBE) [inline HeaderOptionsPanel.h:5]: Wrong copyright info …

File: HeaderOptionsPanel.h [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+Released under the GNU GPLv2, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "Panel.h"
+#include "ScreenManager.h"
+#include "Settings.h"
+
```
Comment (BenBE) [inline HeaderOptionsPanel.h:13]: Double blank line missing …

File: Settings.h [hunk 17]
```diff
@@ -12,20 +12,22 @@ in the source distribution for its full text.
...
 #include "Process.h"
 
 
 #define DEFAULT_DELAY 15
 
 typedef struct {
-   int len;
+   uint8_t len;
    char** names;
    int* modes;
```
Comment (BenBE) [inline Settings.h:24]: Why not use `MeterModeId` from `Meter.h`?

**Review by BenBE — CHANGES_REQUESTED** [review]

File: Settings.h (hunk 17 above)
Comment (cgzones) [reply Settings.h:24]: Cause including *Meter.h* would cause an include loop and one can not forward declare enums.

**Comment by fasterit** [comment]
Please apply the patch from https://github.com/htop-dev/htop/pull/497#discussion_r566235198 for more right-margin-consistency™.

**Comment by cgzones** [comment]
> Please apply the patch from #497 (comment) for more right-margin-consistency™.

Applied.

File: CategoriesPanel.c [hunk 18]
```diff
@@ -34,13 +36,23 @@ static void CategoriesPanel_delete(Object* object) {
...
-   rightMeters->leftNeighbor = leftMeters;
-   Panel* availableMeters = (Panel*) AvailableMetersPanel_new(this->settings, this->header, (Panel*) leftMeters, (Panel*) rightMeters, this->scr, this->pl);
-   ScreenManager_add(this->scr, (Panel*) leftMeters, 20);
-   ScreenManager_add(this->scr, (Panel*) rightMeters, 20);
+   size_t columns = HeaderLayout_getColumns(this->scr->header->headerLayout);
+   MetersPanel** meterPanels = xMallocArray(columns, sizeof(MetersPanel));
+
+   for (size_t i = 0; i < columns; i++) {
+      char titleBuffer[32];
+      xSnprintf(titleBuffer, sizeof(titleBuffer), "%zu. Column", i + 1);
```
Comment (BenBE) [inline CategoriesPanel.c:44]: This probably should avoid the problem of display of ordinal numbers, as the correct counting would probably be "1st column", "2nd column", "3rd column" and so on.
```suggestion
      xSnprintf(titleBuffer, sizeof(titleBuffer), "Column %zu", i + 1);
```

File: Header.c [hunk 19]
```diff
@@ -118,22 +163,27 @@ void Header_setMode(Header* this, int i, MeterModeId mode, int column) {
...
    return Vector_size(meters);
 }
 
-MeterModeId Header_readMeterMode(const Header* this, int i, int column) {
-   const Vector* meters = this->columns[column];
+MeterModeId Header_readMeterMode(const Header* this, int i, unsigned int column) {
+   assert(column < HeaderLayout_getColumns(this->headerLayout));
 
+   const Vector* meters = this->columns[column];
    const Meter* meter = (const Meter*) Vector_get(meters, i);
```
Comment (BenBE) [inline Header.c:187]: Additional asserts may check `meters` and `meter` to be non-NULL for DiD.

File: HeaderLayout.h [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+   [HF_THREE_25_25_50]   = { 3, { 25, 25, 50,  0 }, "3 columns - 25/25/50", },
+   [HF_THREE_25_50_25]   = { 3, { 25, 50, 25,  0 }, "3 columns - 25/50/25", },
+   [HF_THREE_50_25_25]   = { 3, { 50, 25, 25,  0 }, "3 columns - 50/25/25", },
+   [HF_THREE_40_20_40]   = { 3, { 40, 20, 40,  0 }, "3 columns - 40/20/40", },
+   [HF_FOUR_25_25_25_25] = { 4, { 25, 25, 25, 25 }, "4 columns - 25/25/25/25", },
+};
+
+static inline unsigned int HeaderLayout_getColumns(HeaderLayout hLayout) {
+   /* assert the layout is initialized */
+   assert(HeaderLayout_layouts[hLayout].description[0]);
```
Comment (BenBE) [inline HeaderLayout.h:48]: Missing assert that `hLayout` actually is in range … (DiD)

File: HeaderLayout.h [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+   HF_THREE_25_50_25,
+   HF_THREE_50_25_25,
+   HF_THREE_40_20_40,
+   HF_FOUR_25_25_25_25,
+   LAST_HEADER_LAYOUT
+} HeaderLayout;
+
+typedef struct HeaderLayoutImpl_ {
+   unsigned int columns;
+   const unsigned int widths[4];
```
Comment (BenBE) [inline HeaderLayout.h:30]: Why not use `size_t` here?

**Review by BenBE — CHANGES_REQUESTED** [review]

File: Header.c [hunk 22]
```diff
@@ -81,7 +122,9 @@ void Header_writeBackToSettings(const Header* this) {
    }
 }
 
-MeterModeId Header_addMeterByName(Header* this, const char* name, int column) {
+MeterModeId Header_addMeterByName(Header* this, const char* name, unsigned int column) {
+   assert(column < HeaderLayout_getColumns(this->headerLayout));
+
    Vector* meters = this->columns[column];
 
    char* paren = strchr(name, '(');
```
Comment (BenBE) [inline Header.c:130]: Shouldn't it be `meter->mode = mode;` in the `mode` assignment below?

**Review by BenBE — CHANGES_REQUESTED** [review]

File: Header.c [hunk 23]
```diff
@@ -20,51 +21,91 @@ in the source distribution for its full text.
...
+
+   if (newColumns == oldColumns)
+      return;
+
+   if (newColumns > oldColumns) {
+      this->columns = xReallocArray(this->columns, newColumns, sizeof(Vector*));
+      for (size_t i = oldColumns; i < newColumns; i++)
+         this->columns[i] = Vector_new(Class(Meter), true, DEFAULT_SIZE);
+   } else {
+      // move meters from to-deleted columns into last one
```
[the remark this answers is not in the document]
Comment (cgzones) [reply Header.c]: Thanks, fixed

File: Header.c (hunk 22 above)
Comment (cgzones) [reply Header.c]: It was somehow correct, but the returned mode was nowhere used; refactored the function

File: HeaderLayout.h (hunk 21 above)
Comment (cgzones) [reply HeaderLayout.h]: Using `uint8_t` now cause more than 255 header `columns` is unfeasible and `widths` is just a percentage number between 0 and 100.
The access wrapper now returns `size_t` so callers don't need to worry about the actual type.

**Comment by BenBE** [comment]
If I didn't miss anything, there's currently still two review comments unresolved:
1. `ColorPanels.h`: Function signature change of `ColorsPanel_new`
2. `Header.c`: Zero-initialization with `xCalloc` for `Header` instance.

Also please check the latest rebase for any artifacts that might have slipped through.

File: ColorsPanel.h (hunk 8 above)
Comment (fasterit) [reply ColorsPanel.h:24]: Seems not at all, just cleanup while at it

File: Header.c (hunk 9 above)
Comment (fasterit) [reply Header.c:27]: Yes, the xCalloc is better as the struct is not fully initialized in the lines below (pad, height).

**Review by BenBE — APPROVED** [review]

File: Header.c [hunk 24]
```diff
@@ -22,135 +23,156 @@ in the source distribution for its full text.
...
       if (!ok) {
+         char* end, dynamic[32] = {0};
          if (sscanf(paren, "(%30s)", dynamic)) { // DynamicMeter
             if ((end = strrchr(dynamic, ')')) == NULL)
-               return false;    // indicate htoprc parse failure
+               return;    // indicate htoprc parse failure
             *end = '\0';
             if (!DynamicMeter_search(this->pl->dynamicMeters, dynamic, &param))
-               return false;    // indicates name lookup failure
+               return;    // indicates name lookup failure
```
Comment (natoscott) [inline Header.c:93]: This comment doesn't make sense anymore  (no return value but its indicating failure?)  Similarly on line 90, three lines prior.

**Review by natoscott — APPROVED** [review]
Minor comment updates needed I think, otherwise LGTM.

File: Header.c (hunk 24 above)
Comment (BenBE) [reply Header.c:93]: I saw those too yesterday. Will fix them later, unless someone else changes them first … ;-)

Comment (fasterit) [reply Header.c:93]: I think the comments still make sense to explain the error conditions that trigger the (silent) return.

Comment (BenBE) [reply Header.c:93]: What about dropping the `indicates`?

Comment (fasterit) [reply Header.c:93]: yeah, still serves the purpose
